### ISSUE TITLE
Possible race condition in Akka.Cluster

### ISSUE THREAD

**Issue by ismaelhamed** [akkadotnet/akka.net#5435]
**Version Information**
Akka 1.4.29
Akka.Cluster

**Describe the bug**
After upgrading to Akka v1.4.29, I've been hit with this a few times already during testing:

```
[akka://XXXXXX/system/cluster/$a#415417592]: Akka.Actor.ActorInitializationException: Exception during creation ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Akka.Cluster.OnMemberStatusChangedListener.PreStart()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.Create(Exception failure)
   --- End of inner exception stack trace ---
   at Akka.Actor.ActorCell.Create(Exception failure)
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)
```
Looks like this only happens in nodes with a proxy to a sharding, or that's my observation so far. 

And then I can also see this in some of those nodes:

```
System.NotSupportedException: Nobody does not provide
   at Akka.Actor.Nobody.get_Provider()
   at Akka.Actor.Futures.Ask[T](ICanTell self, Func`2 messageFactory, Nullable`1 timeout, CancellationToken cancellationToken)
   at Akka.Actor.Futures.Ask[T](ICanTell self, Object message, Nullable`1 timeout, CancellationToken cancellationToken)
   at Akka.Actor.Futures.Ask[T](ICanTell self, Object message, Nullable`1 timeout)
```
Affected nodes won't recover from this on their own (although the cluster is fully formed and healthy).

I haven't found a way to reproduce it: just restarting nodes or the whole cluster during testing will eventually hit this.

**Comment by Zetanova** [comment]
A view PR's since akka 1.4.21 removed view thread-locks from components and startup 
This made already existing race condition in the startup visible

I fixed already view of them in PR https://github.com/akkadotnet/akka.net/pull/5416
but it did not made it to 1.4.29 

@ismaelhamed Pls try your test with the nightly build.

Yes its very hard to reproduce them.
we already planning to rework the akka startup process

**Comment by Aaronontheweb** [comment]
Looks like this is what introduced those issues https://github.com/akkadotnet/akka.net/pull/5398

**Comment by Aaronontheweb** [comment]
But yes, it would be very helpful to hear if the nightlies, which include https://github.com/akkadotnet/akka.net/pull/5416, alleviate this issue. If that's the case we'll do a release before Christmas.

**Comment by Aaronontheweb** [comment]
What in 1.4.21 would have caused this @Zetanova ? https://github.com/akkadotnet/akka.net/releases/tag/1.4.21 - I was looking through the set of commits and I'm not sure which one you're referring to

**Comment by Zetanova** [comment]
after 1.4.21 we changed `Ask` and `PipeTo` and `Cluster.Get(sys)` got removed from Cluster Components and one more lock.

With 1.4.21 akka-system required at least 4 real threads to startup, 3 locked threads and 1 running.

Cluster components used `Cluster.Get(sys)` that produces a lock because of the `Lay<Extension>`,
but was highly depended on the thread-Id,  after we removed the first lock in the `FutureActor` this got already instable,
because the `ActorCell` could get scheduled on the same thread as the Cluster Extension Thread and throwed the Lazy-Exception

**Comment by Zetanova** [comment]
and its not only in cluster, single system to, the whole startup is/was racy
this includes the akka-logging too.

I thing there currently only one racy startup component left,
If we find and fix it all tests should work too.

Or we refactor the akka-system startup

**Comment by Aaronontheweb** [comment]
I don't recall any of the changes we made to `Ask` or `PipeTo` having these negative side effects (the `PipeTo` changes only happened in 1.4.29 and weren't semantically different from what we had before - just using an `async` local function instead of `ContinueWith`) - but I'll take your word for it about the startup process being racy. I know two years ago we had to remove some of the locking around the logging system because the `ActorSystem` was frequently locking up inside Xamarin environments.

Probably the right way to fix this is to have an `ActorSystem.Create` method that returns `Task<ActorSystem>` and uses `await`s all the way down to start these components but that's a major structural change.

What's the last racy startup component that you're aware of?

**Comment by Arkatufus** [comment]
I can confirm that on dev branch the NRE happens almost all the time if the cluster singleton proxy test is run inside a very CPU limited environment, but not on 1.4.28

How to reproduce:
- In the USERPROFILE directory (C:\Users\[UserName]), create a .wslconfig file:
```
[wsl2]
memory=3GB   # Limits VM memory in WSL 2 up to 3GB
processors=1 # Makes the WSL 2 VM use one virtual processors
```
- Start Linux WSL2, go to the `akka.net/src/contrib/cluster/Akka.Cluster.Tools.Tests` folder
- run `dotnet test --filter "FullyQualifiedName~ClusterSingletonProxySpec"`

**Comment by Zetanova** [comment]
I make now a PR for async startup: https://github.com/akkadotnet/akka.net/pull/5436 as written above

**Comment by Arkatufus** [comment]
Ran the new PR under the same failing condition as the dev branch and it works :)

**Comment by Aaronontheweb** [comment]
Yeah the more I think about this - forcing every single plugin to be completely initialized before the system can do anything is a bad idea and is, as we're seeing, inherently fragile. I don't think https://github.com/akkadotnet/akka.net/pull/5436 is a great solution - it's hacks on top of hacks. I would much rather see an organized "write a spec first, get buy in, implement" approach to solving the problems @Zetanova has raised rather than the piecemeal bits we've merged thus far.

I'm leaning towards reverting #5398 and the parts of #5416 that all touch the startup area and trying this again with a clean slate.

**Comment by Aaronontheweb** [comment]
For the record, I really appreciate @Zetanova's prompt attention on the issue and his work on fixing it. That we can't start Akka.Cluster on a single thread _is_ a legitimate issue but I think we need to do some more on paper design work first.

**Comment by Aaronontheweb** [comment]
Looks like it was only #5398 that needed to be reverted, based on how the tests inside #5437 have been going.

**Comment by ismaelhamed** [comment]
We're reverting to v1.4.28 for now then. Thank you, guys, for jumping on this so promptly!

**Comment by Aaronontheweb** [comment]
Resolved via #5437 - we'll push a 1.4.30 release today that mitigates this issue.
